org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/schema_reference_4CodeAction.java

```diff
@@ -0,0 +1,66 @@
+package org.eclipse.lemminx.extensions.contentmodel.participants.codeactions;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.lemminx.commons.CodeActionFactory;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.extensions.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.utils.StringUtils;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * Code Action that creates a schema file referenced by xsi:noNamespaceSchemaLocation if it is missing
+ */
+public class schema_reference_4CodeAction implements ICodeActionParticipant {
+
+	private static final Pattern PATH_FINDING_REGEX = Pattern.compile("[^']+'file://(.+)',.*");
+
+	@Override
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
+			SharedSettings sharedSettings, IComponentProvider componentProvider) {
+
+		String missingFilePath = getPathFromDiagnostic(diagnostic);
```

Comment:
Maybe we can save Workspace edit capabilities in `sharedSettings` here: https://github.com/eclipse/lemminx/blob/59c04ef37c7959d95044f61564939c2f139b2d06/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLTextDocumentService.java#L159-L184

and then in `doCodeAction()`, we can check for the capabilities stored in `sharedSettings`?

Reply:
Fred suggested just not registering `schema_reference4CodeAction` if the client doesn't support the `CreateFile` `ResourceOperation`, but since `SharedSettings` aren't available in the registry method, I might have to either refactor a lot of things or go this route.

Reply:
I pushed a revision that implements this the way you suggested.